### DISCUSSION TITLE
How do I get a template class name inside an extending class method?

### DISCUSSION THREAD

**Discussion by designermonkey** [vimeo/psalm#10133] [General]
Is there any way, or with any utilities, to get a class name used as a generic template value inside my method code?

For example, in this Java code:
```java
public abstract class Application<T extends Configuration> {
  public Class<T> getConfigurationClass() {
      return Generics.getTypeParameter(getClass(), Configuration.class);
  }
}
```

You can see that it's trivial to get the name of the class `T` extending `Configuration` as generics are a language feature.

So far, I know that in PHP:
```php
/**
 * @template TConfiguration of Configuration
 */
abstract class Application
{
    /**
     * @psalm-return class-string<TConfiguration>
     */
    public function getConfigurationClass(): string
    {
        // what goes here?
    }
}
```

My question is, what should I be doing in the `getConfigurationClass` method? Are there any utilities in Psalm that allow me to do this, or do I have to do `ReflectionClass` work to get the doc strings and regex parse them?

**Comment by weirdan** [comment]
There's no runtime support for generics in PHP, with or without Psalm. While you probably can parse the docblock and figure out the generic parameter, that sounds like a lot of work that could be avoided. You could just let the extending class provide the type for you, e.g.:

```php
/**
 * @template TConfiguration of Configuration
 */
abstract class Application
{
    /**
     * @psalm-return class-string<TConfiguration>
     */
    abstract public function getConfigurationClass(): string;
}

// somewhere in the client code

class CustomerPortalConfiguration extends Configuration {}

/** @extends Application<CustomerPortalConfiguration> */
class CustomerPortalApplication extends Application 
{
    public function getConfiguration(): string 
    {
        return CustomerPortalConfiguration::class;
    }
}
```

**Reply by designermonkey** [reply]
I know about lack of generics support, and I did think I may have to use reflection. I like your suggestion, though I was hoping there was a way I could do it without having to manually code the string into a method body, i.e. use a function in the parent abstract class per the example I gave.

Are there utilities that are easy to use from Psalm to parse the docblock parts, or do I have to reinvent the wheel so to speak?

**Reply by weirdan** [reply]
Not from Psalm, but PHPStan has a docblock parser that supports most of the Psalm's features.

See https://github.com/phpstan/phpdoc-parser

**Reply by designermonkey** [reply]
I'll take a look! Thanks @weirdan
